File: src/main/java/net/ballmerlabs/scatterbrainsdk/Identity.java
package net.ballmerlabs.scatterbrainsdk;

import android.os.Parcel;
import android.os.Parcelable;

import androidx.annotation.NonNull;

import java.util.AbstractMap;
import java.util.HashMap;
import java.util.Map;

public class Identity implements Parcelable {

    protected final Map<String, byte[]> mPubKeymap;
    protected final byte[] mScatterbrainPubKey;
    protected final String givenname;
    protected final byte[] sig;
    protected final String fingerprint;
    protected boolean hasPrivateKey;

    protected Identity(
            Map<String, byte[]> map, 
            byte[] pub, 
            String name, 
            byte[] sig, 
            String fingerprint, 
            boolean hasPrivateKey
    ) {
        this.mPubKeymap = map;
        this.mScatterbrainPubKey = pub;
        this.givenname = name;
        this.sig = sig;
        this.fingerprint = fingerprint;
        this.hasPrivateKey = hasPrivateKey;
    }

    @FunctionalInterface
    private interface ParcelWriter<T> {
        void writeToParcel(@NonNull final T value,
                           @NonNull final Parcel parcel, final int flags);
    }

    @FunctionalInterface
    private interface ParcelReader<T> {
        T readFromParcel(@NonNull final Parcel parcel);
    }

    private static <K, V> void writeParcelableMap(
            @NonNull final Map<K, V> map,
            @NonNull final Parcel parcel,
            final int flags,
            @NonNull final ParcelWriter<Map.Entry<K, V>> parcelWriter) {
        parcel.writeInt(map.size());

        for (final Map.Entry<K, V> e : map.entrySet()) {
            parcelWriter.writeToParcel(e, parcel, flags);
        }
    }

    private static <K, V> Map<K, V> readParcelableMap(
            @NonNull final Parcel parcel,
            @NonNull final ParcelReader<Map.Entry<K, V>> parcelReader) {
        int size = parcel.readInt();
        final Map<K, V> map = new HashMap<>(size);

        for (int i = 0; i < size; i++) {
            final Map.Entry<K, V> value = parcelReader.readFromParcel(parcel);
            map.put(value.getKey(), value.getValue());
        }
        return map;
    }

    protected Identity(Parcel in) {
        mPubKeymap = readParcelableMap(in, parcel -> {
            final int len = parcel.readInt();
            final byte[] key = new byte[len];
            parcel.readByteArray(key);
            return new AbstractMap.SimpleEntry<>(parcel.readString(), key);
        });
        mScatterbrainPubKey = mPubKeymap.get(ScatterbrainApi.PROTOBUF_PRIVKEY_KEY);
        givenname = in.readString();
        sig = new byte[in.readInt()];
        in.readByteArray(sig);
        fingerprint = in.readString();
        hasPrivateKey = hasKey(in.readByte());
    }

    public static final Creator<Identity> CREATOR = new Creator<Identity>() {
        @Override
        public Identity createFromParcel(Parcel in) {
            return new Identity(in);
        }

        @Override
        public Identity[] newArray(int size) {
            return new Identity[size];
        }
    };

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel parcel, int i) {
        writeParcelableMap(mPubKeymap, parcel, i, (mapentry, p, __) -> {
            p.writeInt(mapentry.getValue().length);
            p.writeByteArray(mapentry.getValue());
            p.writeString(mapentry.getKey());
        });
        parcel.writeString(givenname);
        parcel.writeInt(sig.length);
        parcel.writeByteArray(sig);
        parcel.writeString(fingerprint);
        parcel.writeByte(hasKey(hasPrivateKey));
    }

    public Map<String, byte[]> getmPubKeymap() {
        return mPubKeymap;
    }

    public byte[] getmScatterbrainPubKey() {
        return mScatterbrainPubKey;
    }

    public String getGivenname() {
        return givenname;
    }

    public byte[] getSig() {
        return sig;
    }

    public String getFingerprint() {
        return fingerprint;
    }
    
    public boolean hasPrivateKey() {
        return hasPrivateKey;
    }

    protected void setHasPrivateKey(boolean key) {
        this.hasPrivateKey = key;
    }


    private static boolean hasKey(int val) {
        if (val == 0) {
            return true;
        } else {
            return false;
        }
    }

    private static byte hasKey(boolean val) {
        if (val) {
            return 0;
        } else {
            return 1;
        }
    }
}

File: scatterbrainSDK/src/main/java/net/ballmerlabs/scatterbrainsdk/Identity.kt
package net.ballmerlabs.scatterbrainsdk

import android.os.Parcel
import android.os.ParcelUuid
import android.os.Parcelable
import androidx.compose.runtime.Stable
import kotlinx.collections.immutable.ImmutableMap
import kotlinx.collections.immutable.toImmutableMap
import net.ballmerlabs.scatterbrainsdk.internal.b64
import net.ballmerlabs.scatterbrainsdk.internal.readBool
import net.ballmerlabs.scatterbrainsdk.internal.readParcelableMap
import net.ballmerlabs.scatterbrainsdk.internal.writeBool
import net.ballmerlabs.scatterbrainsdk.internal.writeParcelableMap
import java.util.AbstractMap
import java.util.UUID

fun parcelArray(parcel: Parcel): ByteArray {
    val s = ByteArray(parcel.readInt())
    parcel.readByteArray(s)
    return s
}

const val PROTOBUF_PRIVKEY_KEY = "scatterbrain"

/**
 * A handle to a cryptographic identity stored in the Scatterbrain
 * router. This class contains all identity metadata except for the
 * private key (for security reasons). This class has no public constructor
 * and is only returned by Scatterbrain api functions
 *
 * @property publicKey ed25519 public key used by Scatterbrain
 * @property extraKeys additional user defined keys or metadata
 * @property sig ed25519 signature for this identity
 * @property fingerprint unique identifier for this identity
 * @property name user-defined name
 * @property isOwned true if this identity has a private key
 */
@Stable
data class Identity(
    @Stable
    val extraKeys: ImmutableMap<String, ByteArray>,
    @Stable
    val publicKey: ByteArray = extraKeys[PROTOBUF_PRIVKEY_KEY]!!,
    @Stable
    val name: String,
    @Stable
    val sig: ByteArray,
    @Stable
    val fingerprint: UUID,
    @Stable
    val isOwned: Boolean,
    @Stable
    val frozen: Boolean,
) : Parcelable {
    constructor(inParcel: Parcel) : this(
        extraKeys = readParcelableMap(inParcel) { parcel ->
            val len = parcel.readInt()
            val key = ByteArray(len)
            parcel.readByteArray(key)
            AbstractMap.SimpleEntry(parcel.readString()!!, key)
        }.toImmutableMap(),
        name = inParcel.readString()!!,
        sig = parcelArray(inParcel),
        fingerprint = inParcel.readParcelable<ParcelUuid>(ParcelUuid::class.java.classLoader)!!.uuid!!,
        isOwned = hasKey(inParcel.readByte().toInt()),
        frozen = inParcel.readBool()
    )

    @Stable
    override fun describeContents(): Int {
        return 0
    }

    @Stable
    override fun writeToParcel(parcel: Parcel, i: Int) {
        writeParcelableMap(extraKeys, parcel, i) { mapentry, p, _ ->
            p.writeInt(mapentry.value.size)
            p.writeByteArray(mapentry.value)
            p.writeString(mapentry.key)
        }
        parcel.writeString(name)
        parcel.writeInt(sig.size)
        parcel.writeByteArray(sig)
        parcel.writeParcelable(ParcelUuid(fingerprint), i)
        parcel.writeByte(hasKey(isOwned))
        parcel.writeBool(frozen)
    }

    @Stable
    override fun toString(): String {
        var id = "Identity(" +
                "sig=${this.sig.b64()}\n" +
                "pubkey=${this.publicKey.b64()}\n" +
                "name=${this.name}\n" +
                "isOwned=${this.isOwned}\n" +
                "fingerprint=${this.fingerprint}"

        for ((k, v) in extraKeys) {
            id += "     ($k, ${v.b64()})\n"
        }
        return id
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true

        other as Identity

        if (extraKeys.keys.any { k ->
                !other.extraKeys.containsKey(k) || !other.extraKeys[k].contentEquals(
                    extraKeys[k]
                )
            }) return false

        if (!publicKey.contentEquals(other.publicKey)) return false
        if (name != other.name) return false
        if (!sig.contentEquals(other.sig)) return false
        if (fingerprint != other.fingerprint) return false
        if (isOwned != other.isOwned) return false

        return true
    }

    override fun hashCode(): Int {
        var result = 0
        extraKeys.forEach { (k, v) ->
            result = 31 * result + k.hashCode()
            result = 31 * result + v.contentHashCode()
        }
        result = 31 * result + publicKey.contentHashCode()
        result = 31 * result + name.hashCode()
        result = 31 * result + sig.contentHashCode()
        result = 31 * result + fingerprint.hashCode()
        result = 31 * result + isOwned.hashCode()
        return result
    }

    companion object {
        @JvmField
        val CREATOR: Parcelable.Creator<Identity> = object : Parcelable.Creator<Identity> {
            override fun createFromParcel(`in`: Parcel): Identity {
                return Identity(`in`)
            }

            override fun newArray(size: Int): Array<Identity?> {
                return arrayOfNulls(size)
            }
        }

        fun hasKey(`val`: Int): Boolean {
            return `val` == 0
        }

        fun hasKey(`val`: Boolean): Byte {
            return if (`val`) {
                0
            } else {
                1
            }
        }
    }
}